docs/bos/tutorial/queryapi-ws.md

```diff
@@ -0,0 +1,344 @@
+  soundEffect:
+    "https://bafybeic7uvzmhuwjficgctpleov5i43rteavwmktyyjrauwi346ntgja4a.ipfs.nftstorage.link/",
+});
+
+const widgetActivitySubscription = `
+  subscription IndexerQuery {
+    roshaan_near_widget_activity_feed_widget_activity(
+      order_by: {block_timestamp: desc}
+      limit: ${LIMIT}
+    ) {
+      account_id
+      block_height
+      block_timestamp
+      id
+      receipt_id
+      widget_name
+    }
+  }
+`;
+
+const subscriptionWidgetActivity = {
+  type: "start",
+  id: "widgetActivity", // You can use any unique identifier
+  payload: {
+    operationName: "IndexerQuery",
+    query: widgetActivitySubscription,
+    variables: {},
+  },
+};
+function processWidgetActivity(activity) {
+  return { ...activity };
+}
+function startWebSocketWidgetActivity(processWidgetActivities) {
+  let ws = State.get().ws_widgetActivity;
+
+  if (ws) {
+    ws.close();
+    return;
+  }
+
+  ws = new WebSocket(`wss://${GRAPHQL_ENDPOINT}/v1/graphql`, "graphql-ws");
+
+  ws.onopen = () => {
+    console.log(`Connection to WS has been established`);
+    ws.send(
+      JSON.stringify({
+        type: "connection_init",
+        payload: {
+          headers: {
+            "Content-Type": "application/json",
+            "Hasura-Client-Name": "hasura-console",
+            "x-hasura-role": "roshaan_near",
+          },
+          lazy: true,
+        },
+      })
+    );
+
+    setTimeout(() => ws.send(JSON.stringify(subscriptionWidgetActivity)), 50);
+  };
+
+  ws.onclose = () => {
+    State.update({ ws_widgetActivity: null });
+    console.log(`WS Connection has been closed`);
+  };
+
+  ws.onmessage = (e) => {
+    const data = JSON.parse(e.data);
+    console.log("received data", data);
+    if (data.type === "data" && data.id === "widgetActivity") {
+      processWidgetActivities(data.payload.data);
+    }
+  };
+
+  ws.onerror = (err) => {
+    State.update({ ws_widgetActivity: null });
+    console.log("WebSocket error", err);
+  };
+
+  State.update({ ws_widgetActivity: ws });
+}
+```
+
+---
+
+### Processing
+
+This is the JS function that process the incoming widget activities generated by the QueryAPI indexer, allowing the BOS component to create a feed based on the blockchain's widget activity:
+
+
+:::tip
+
+You can fork the [Widget Activity Feed source code](https://near.org#/near/widget/ComponentDetailsPage?src=roshaan.near/widget/query-api-widget-feed) and build your own BOS component.
+
+:::
+
+```js
+function processWidgetActivities(incoming_data) {
+  let incoming_widgetActivities =
+    incoming_data.roshaan_near_widget_activity_feed_widget_activity.flatMap(
+      processWidgetActivity
+    );
+  const newActivities = [
+    ...incoming_widgetActivities.filter((activity) => {
+      return (
+        state.widgetActivities.length == 0 ||
+        activity.block_timestamp > state.widgetActivities[0].block_timestamp
+      );
+    }),
+  ];
+  if (newActivities.length > 0 && state.widgetActivities.length > 0) {
```

Comment:
You may just want to get rid of anything to do with sounds as it does not work haha

Reply:
removed 👍